File: src/Baekjoon/p10866.java
package Baekjoon;

import java.io.*;

public class p10866 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder bw = new StringBuilder();

        int N = Integer.parseInt(br.readLine());
        Deque deque = new Deque();

        for (int i = 0; i < N; i++) {
            String[] command = br.readLine().split(" ");

            if (command[0].equals("push_front")) {
                int x = Integer.parseInt(command[1]);
                deque.pushFront(x);
            }
            else if (command[0].equals("push_back")) {
                int x = Integer.parseInt(command[1]);
                deque.pushBack(x);
            }
            else if (command[0].equals("pop_front")) {
                bw.append(deque.popFront()).append('\n');
            }
            else if (command[0].equals("pop_back")) {
                bw.append(deque.popBack()).append('\n');
            }
            else if (command[0].equals("size")) {
                bw.append(deque.size()).append('\n');
            }
            else if (command[0].equals("empty")) {
                bw.append(deque.empty()).append('\n');
            }
            else if (command[0].equals("front")) {
                bw.append(deque.front()).append('\n');
            }
            else if (command[0].equals("back")) {
                bw.append(deque.back()).append('\n');
            }
        }

        System.out.print(bw);
    }

    private static class Deque {
        private int[] deque;
        private int frontIndex;
        private int backIndex;

        public Deque() {
            deque = new int[20000];
            frontIndex = 10000;
            backIndex = 9999;
        }

        public void pushFront(int x) {
            // 앞으로 한 칸 이동 (-1)
            deque[--frontIndex] = x;
        }

        public void pushBack(int x) {
            // 뒤로 한 칸 이동 (+1)
            deque[++backIndex] = x;
        }

        public int popFront() {
            if (empty() == 1) {
                return -1;
            }
            // 뒤로 한 칸 이동 (+1)
            return deque[frontIndex++];
        }

        public int popBack() {
            if (empty() == 1) {
                return -1;
            }
            // 앞으로 한 칸 이동 (-1)
            return deque[backIndex--];
        }

        public int size() {
            return backIndex - frontIndex + 1;
        }

        public int empty() {
            if (backIndex < frontIndex) {
                return 1;
            }
            return 0;
        }

        public int front() {
            if (empty() == 1) {
                return -1;
            }
            return deque[frontIndex];
        }

        public int back() {
            if (empty() == 1) {
                return -1;
            }
            return deque[backIndex];
        }
    }
}

File: src/Baekjoon/p10866.cpp
#include <iostream>
#include <string>

using namespace std;


// 전역변수 선언
int N, X;

// Deque 클래스 구현
class Deque {
    private:
        int container[20000];
        int frontIdx = 9999;
        int backIdx = 10000;
    
    public:
        void push_front(int X) {
            container[frontIdx--] = X;
        }

        void push_back(int X) {
            container[backIdx++] = X;
        }

        int pop_front() {
            if (backIdx - frontIdx == 1) return -1;
            else return container[++frontIdx];
        }

        int pop_back() {
            if (backIdx - frontIdx == 1) return -1;
            else return container[--backIdx];
        }

        int size() {
            return backIdx - frontIdx - 1;
        }

        int empty() {
            if (backIdx - frontIdx == 1) return 1;
            else return 0;
        }

        int front() {
            if (backIdx - frontIdx == 1) return -1;
            else return container[frontIdx + 1];
        }

        int back() {
            if (backIdx - frontIdx == 1) return -1;
            else return container[backIdx - 1];
        }
};

int main(void) {
    // 0. 입출력 시간 줄이기
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    // 1. 변수 선언
    string cmd = "";
    Deque deque = Deque();

    // 2. 입력받아서 명령 처리
    cin >> N;

    for (int i = 0; i < N; i++) {
        cin >> cmd;

        if (cmd == "push_front") {
            cin >> X;
            deque.push_front(X);
        }
        else if (cmd == "push_back") {
            cin >> X;
            deque.push_back(X);
        }
        else if (cmd == "pop_front") {
            cout << deque.pop_front() << '\n';
        } 
        else if (cmd == "pop_back") {
            cout << deque.pop_back() << '\n';
        }
        else if (cmd == "size") {
            cout << deque.size() << '\n';
        }
        else if (cmd == "empty") {
            cout << deque.empty() << '\n';
        }
        else if (cmd == "front") {
            cout << deque.front() << '\n';
        }
        else if (cmd == "back") {
            cout << deque.back() << '\n';
        }
    }
}